src/tests/JIT/opt/InstructionCombining/NegMulOrDivToConst.cs

```diff
@@ -166,7 +169,12 @@ static int CheckNegMul()
         }
 
         [MethodImpl(MethodImplOptions.NoInlining)]
-        static int NegMul7(int a) => -(a * 7);
+        static int NegMul7(int a)
+        {
+            //ARM64-FULL-LINE: movn {{w[0-9]+}}, #6
+            //ARM64-FULL-LINE-NEXT: mul {{w[0-9]+}}, {{w[0-9]+}}, {{w[0-9]+}}
+            return -(a * 7);
```

Comment:
can you add few examples where non-integral multiplication or where `gtOverFlow()` is set?

Reply:
This doesn't look right. If I see the asmdiff, I clearly see `mneg` instruction is getting generated.

![image](https://user-images.githubusercontent.com/12488060/211892157-bf765726-f0c5-4731-9733-9a708c65b262.png)

Reply:
Also, for `CheckedMulNegIntMin` and `NegMulIntMaxValue`, my understanding is that `mneg` shouldn't happen because of it could have `GTF_OVERFLOW`? Can you please double check?

![image](https://user-images.githubusercontent.com/12488060/211892882-77a301ba-8469-4f2b-b5bc-a22eff8a50e7.png)

Reply:
Hi Kunal,

When I compiled `MulNeg7()` method separately using the following flags, I still get a combination of movn + mul. 

```
COMPlus_JitDiffableDasm=1 COMPlus_JitDisasm="MulNeg7" DOTNET_TieredCompilation=0 ~/dotnet/runtime/artifacts/tests/coreclr/Linux.arm64.Checked/Tests/Core_Root/corerun ~/dotnet/mneg/bin/Release/net7.0/mneg.dll 
```


<details open>
<summary>Assembly</summary>

```
; Assembly listing for method Demo:MulNeg7(int):int
; Emitting BLENDED_CODE for generic ARM64 CPU - Unix
; optimized code
; fp based frame
; partially interruptible
; No PGO data
; Final local variable assignments
;
;  V00 arg0         [V00,T00] (  3,  3   )     int  ->   x0         single-def
;# V01 OutArgs      [V01    ] (  1,  1   )  lclBlk ( 0) [sp+00H]   "OutgoingArgSpace"
;
; Lcl frame size = 0

G_M25623_IG01:
            stp     fp, lr, [sp, #-0x10]!
            mov     fp, sp
						;; size=8 bbWeight=1 PerfScore 1.50
G_M25623_IG02:
            movn    w1, #6
            mul     w0, w0, w1
						;; size=8 bbWeight=1 PerfScore 2.50
G_M25623_IG03:
            ldp     fp, lr, [sp], #0x10
            ret     lr
						;; size=8 bbWeight=1 PerfScore 2.00

; Total bytes of code 24, prolog size 8, PerfScore 8.40, instruction count 6, allocated bytes for code 24 (MethodHash=4a419be8) for method Demo:MulNeg7(int):int
```
</details>

In the above assembly, I don't get `; MinOpts Code` and `; Tier-0 compilation`. Are there any flags I'm missing while running this?
What is MinOpts? I looked for MinOpts and I found `DOTNET_JITMinOpts=1` flag. However, couldn't get the assembly that you are seeing.
I'm using ubuntu 20.04 on an N1 machine. We can remove the asm checks if they are not deterministic on the different machines.

Reply:
> can you add few examples where non-integral multiplication or where `gtOverFlow()` is set?

The `CheckedMul*()` tests are explicitly checking for the overflow where we don't emit mneg. However, not sure how to force the `GTF_OVERFLOW` flag using non-integral multiplication. Could you suggest, maybe an example, please?

Reply:
> Are there any flags I'm missing while running this?

Actually, I just realized that the diffs I showed was for `Tier-0 compilation`. You get that only when `DOTNET_TieredCompilation` **is not set**. If it odd that in tier0 we get `mneg`, but in optimized code we don't. Can you check why that is the case?

> `GTF_OVERFLOW`

That's fine. We are good with the coverage.

Reply:
Seems for `NegMul7()`, the less optimal Tier-0 compilation mode generated the `movn+mul+neg` sequence (latency 1+2+1 on N1). This got optimised by the mneg changes with `movn+mneg` (latency 1+2).
However, with the optimised version of `NegMul7()` gave us `movn+mul` (latency 1+2) which is equivalent to what we achieved with mneg. We shouldn't see any performance difference with either of the sequence.

For `NegMulIntMaxValue()`, the `-(a * int.MaxValue)` operation got optimised as `-int.MaxValue * a` => `mov+mul` (latency 1 + 2).
For `NegMulIntMinValue()`, the `-(a * int.MinValue)` operation got optimised as `-(-a * int.MaxValue)`  => `neg+lsl+neg`  (Latency 1+1+1).
Mneg operates on register operands so we would need to emit a mov to populate the constant value (min/max int) in a register before emitting meng. This would give us `mov+mneg` with the same latency(1+2).

The numeric simplifications while optimising `NegMulIntM*Value` are updating/optimising the `mul+neg` sequence that mneg requires. I'm not sure where exactly these simplifications occur. We may end up in emitting one instruction less for the `NegMulIntMinValue()` scenario but mneg may not give any performance improvements in these scenarios.

Reply:
> We shouldn't see any performance difference with either of the sequence.

Sounds good.